phper/src/types.rs

```diff
@@ -10,162 +10,283 @@
-    pub const LONG: TypeInfo = TypeInfo::from_raw(IS_LONG);
-    pub const NULL: TypeInfo = TypeInfo::from_raw(IS_NULL);
-    pub const OBJECT: TypeInfo = TypeInfo::from_raw(IS_OBJECT);
-    pub const RESOURCE: TypeInfo = TypeInfo::from_raw(IS_RESOURCE);
-    pub const STRING: TypeInfo = TypeInfo::from_raw(IS_STRING);
-    pub const UNDEF: TypeInfo = TypeInfo::from_raw(IS_UNDEF);
+    pub const ARRAY: TypeInfo = TypeInfo::from_type_mask(IS_ARRAY);
+    pub const BOOL: TypeInfo = TypeInfo::from_type_mask(_IS_BOOL);
+    pub const DOUBLE: TypeInfo = TypeInfo::from_type_mask(IS_DOUBLE);
+    pub const FALSE: TypeInfo = TypeInfo::from_type_mask(IS_FALSE);
+    pub const LONG: TypeInfo = TypeInfo::from_type_mask(IS_LONG);
+    pub const MIXED: TypeInfo = TypeInfo::from_type_mask(IS_MIXED);
+    pub const NEVER: TypeInfo = TypeInfo::from_type_mask(IS_NEVER);
+    pub const NULL: TypeInfo = TypeInfo::from_type_mask(IS_NULL);
+    pub const OBJECT: TypeInfo = TypeInfo::from_type_mask(IS_OBJECT);
+    pub const RESOURCE: TypeInfo = TypeInfo::from_type_mask(IS_RESOURCE);
+    pub const STRING: TypeInfo = TypeInfo::from_type_mask(IS_STRING);
+    pub const TRUE: TypeInfo = TypeInfo::from_type_mask(IS_TRUE);
+    pub const UNDEF: TypeInfo = TypeInfo::from_type_mask(IS_UNDEF);
+    pub const VOID: TypeInfo = TypeInfo::from_type_mask(IS_VOID);
+    pub const CALLABLE: TypeInfo = TypeInfo::from_type_mask(IS_CALLABLE);
+    pub const ITERABLE: TypeInfo = TypeInfo::from_type_mask(IS_ITERABLE);
 }
 
 impl TypeInfo {
-    pub const fn from_raw(t: u32) -> Self {
-        Self { t }
+    pub fn from_classname(classname: &str) -> Self {
+        Self {
+            inner: ZendType {
+                ptr: std::ffi::CString::new(classname)
+                    .expect("CString::new failed")
+                    .into_raw() as *mut std::ffi::c_void,
+                type_mask: _ZEND_TYPE_NAME_BIT,
+            },
+        }
+    }
+
+    pub fn varadic(&mut self) -> &Self {
+        self.inner.type_mask |= _ZEND_IS_VARIADIC_BIT;
+
+        self
+    }
+
+    pub const fn from_type_mask(mask: u32) -> Self {
+        Self {
+            inner: ZendType {
+                ptr: std::ptr::null::<std::ffi::c_void>() as *mut std::ffi::c_void,
+                type_mask: mask,
+            },
+        }
+    }
+
+    pub fn from_zval(zval: &ZVal) -> Self {
+        let mask = unsafe { phper_z_type_info_p(zval.as_ptr()) };
+        if (mask & IS_OBJECT) != 0 {
+            let obj = unsafe {
+                let ptr = phper_z_obj_p(zval.as_ptr());
+
+                ZObj::from_mut_ptr(ptr)
+            };
+
+            let name = obj.get_class().get_name().as_ptr();
+
+            Self {
+                inner: ZendType {
+                    ptr: name as *mut std::ffi::c_void,
+                    type_mask: _ZEND_TYPE_NAME_BIT,
+                },
+            }
+        } else {
+            Self::from_type_mask(mask)
+        }
+    }
+
```

Comment:
this is currently broken, `hello` example panics.

can't figure it out, but i'm guessing i need some sleep :) will give it another try later.